backend/GameManager.py

```diff
@@ -58,5 +82,173 @@ def get_pieces_left(self):
         }
 
     def end_turn(self):
-        self.turn = Turn.swap_turn(self.turn)
+        self.turn = Color.swap_turn(self.turn)
         return 1
+
+    def is_adjacent(self, current_row, current_col, target_row, target_col):
+        """ Check if the target position is adjacent to the current position """
+        return (target_row, target_col) in self.board.adjacent_positions_map.get((current_row, current_col), [])
+
+    def is_empty(self, row, col):
+        """ Check if the cell at (row, col) is empty """
+        return self.board.board[row][col] is None
+
+    def is_adjacent_and_empty(self, current_row, current_col, target_row, target_col):
+        """ Check if the target cell is adjacent to the current cell and is empty """
+        if self.is_adjacent(current_row, current_col, target_row, target_col):
+            if (target_row, target_col) in self.open_moves:
+                return True
+        return False
+
+    def find_empty_adjacents(self, row, col):
+        empty_adjacent_positions = []
+        all_adjacent_positions = self.board.adjacent_positions_map[(row, col)]
+
+        for row, col in all_adjacent_positions:
+            position_state = self.board.check_position(row, col)
+            if position_state == Color.EMPTY or position_state == Color.VOID:
+                empty_adjacent_positions.append((row, col))
+
+        return empty_adjacent_positions
+
+    def can_fly(self, player):
+        can_fly = False
+        if player == Color.WHITE and len(self.pieces.white_pieces) <= 3:
+            can_fly = True
+        elif player == Color.BLACK and len(self.pieces.black_pieces) <= 3:
+            can_fly = True
+        return can_fly
+
+    '''
+    the select_piece func should return the available empty adjacent positions to where the piece can be moved 
+    or throw exception if it should be selectable
+    '''
+    def select_piece(self, row, col):
+        if not self.placement_complete():
+            raise Exception("SelectionError -- Cannot select pieces during placement phase")
```

Comment:
Should push a message to the GUI for this.

Reply:
I think we should maybe table that for next sprint. It's not ideal, but we've got enough basic functionality that doesn't work that I'd rather not handle it a day before the due date.

Reply:
okay so currently nothing will happen if opponent select piece during 1st phase may be we can log this into frontend at some point, but in the games i played so far its the same we implemented